Export C functions for other languages

kabili207 opened the pull request:
## Description:
This change exports the necessary decoder functions in C-style so they can be called by other languages. It's heavily based on the Python version already included.

Tested successfully with C# and Go

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] I accept the [DCO](https://github.com/theengs/decoder/blob/development/docs/participate/development.md#developer-certificate-of-origin).

kabili207 committed:
Export C functions for other languages

koenvervloesem commented:
This is interesting. Is it also possible to include sample code for C# and/or Go, like the Python example in the examples directory?

kabili207 commented:
I've added an example for Go, as well as squashed and rebased my commits.

I won't have access to the windows machine for a few days, so a proper C# example will have to wait, but the function declarations below should be the bare minimum to work with the library. A proper C# example would wrap these inside a class but, as I said, no access to windows at the moment.

```csharp
[DllImport("decoder.dll")]
public static extern IntPtr Theengs_NewDecoder();

[DllImport("decoder.dll")]
public static extern void Theengs_DestroyDecoder(IntPtr pointer);

[DllImport("decoder.dll")]
public static extern string Theengs_DecodeBLE(IntPtr pointer, string json_data);

[DllImport("decoder.dll")]
public static extern string Theengs_GetProperties(IntPtr pointer, string model_id);

[DllImport("decoder.dll")]
public static extern string Theengs_GetAttribute(IntPtr pointer, string model_id, string attribute);
```

1technophile commented:
That's great thanks!

DigiH commented:
Thanks @kabili207 !

Do you want to wait for being back on your Windows machine to add a C# example example to this PR, or do you want to do this later with a separate PR?

kabili207 commented:
I can make a separate pull request for the C# example
